CASSANDRA-19946 Uses ParameterizedClass for all auth-related implementations

tiagomlalves opened the pull request:
Uses ParameterizedClass for IAuthorizer, INetworkAuthorizer, and IRoleManager implementations enabling custom implementations to be configurable and read those configurations from cassandra.yaml.

Opportunistically makes consistent the initialization and setting of all auth-related implementations.

---

Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

tiagomlalves committed:
CASSANDRA-19946 Uses ParameterizedClass for all auth-related implementations

Uses ParameterizedClass for IAuthorizer, INetworkAuthorizer, and
IRoleManager implementations enabling those to be configurable from
parameters specified directly in cassandra.yaml.

Opportunistically makes consistent the initialization and setting of
all auth-related implementations and removes code duplication.

Notes:
 * IInternodeAuthenticator implementations are expected to be set as
   default directly in DatabaseDescriptor instead of done via
   AuthConfig.applyAuth(). This is assumed in tests and client logic.
   For instance ReconnectableSnitchHelperTest fetches this authenticator
   before calling daemonInitialize(). Also, BulkLoader fetches this
   directly when creating an outbound connection.
 * Changing this behavior in BulkLoader causes a cascade of changes.
   First, one would need to add AuthConfig.applyAuth() directly in
   clientInitialize() and then would need to enable all the additional
   classes that this change causes. Long term this might be justified
   as if we're depending on the auth logic to be properly set. However,
   in the context of this change it was decided to postpone further
   changes.

smiklosovic commented on src/java/org/apache/cassandra/auth/AuthConfig.java:
```diff
@@ -133,4 +124,21 @@ public static void applyAuth()
+        if (authCls != null && authCls.class_name != null)
+        {
+            String authPackage = AuthConfig.class.getPackage().getName();
+            return ParameterizedClass.newInstance(authCls, Arrays.asList("", authPackage));
```
@tiagomlalves so what happens when authCls is in FQCN form?

tiagomlalves replied:
@smiklosovic thanks for looking at this.

FQCN form is supported by `ParameterizedClass` by specifying an empty search package (first element of the array). This should be required when specifying a own implementation of authentication / authorization classes.

smiklosovic replied:
@tiagomlalves btw I think that it would be nicer if `newInstance` just took `authPackage` and empty string would be covered automatically internally. `Array.asList` might be also changed to `List.of()`. Or even better, `newInstance(authCls, authPackage)` might be like this `newInstance(String clazz, String... packages)` so if there are no packages and you call it like this `newInstance(clazz)` then package will be `""` only hence `clazz` would need to be FQCN which makes total sense to me. More packages would be like `newInstance(clazz, pkg1, pkg2)`

This stuff is often used in integrations so I would keep the original version as well and it would translate to this new form internally.

smiklosovic commented:
@tiagomlalves we should reflect in cassandra.yaml for respective configurations that putting there parameters is also possible.

tiagomlalves commented:
> @tiagomlalves we should reflect in cassandra.yaml for respective configurations that putting there parameters is also possible.

Good point. These changes should be done only in `cassandra_latest.yaml`, right?

smiklosovic commented:
@tiagomlalves no, in main cassandra.yaml for sure, _latest.yaml is also welcome.

smiklosovic reviewed (changes requested):
please apply the suggestions

smiklosovic commented on conf/cassandra.yaml:
```diff
@@ -157,83 +157,105 @@ auto_hints_cleanup_enabled: false
 # - CassandraCIDRAuthorizer stores user's CIDR permissions in system_auth.cidr_permissions table. Please
 #   increase system_auth keyspace replication factor if you use this authorizer, otherwise any changes to
 #   system_auth tables being used by this feature may be lost when a host goes down.
+#   Below parameters are used only when CIDR authorizer is enabled
```
@tiagomlalves  well, this pattern you use here is not what we exactly want. The issue is with the ordering. How it is supposed to work is that as a reader reads the configuration file, he will just un-comment respective configuration property _in place_. 

In other words, we do not want a user to copy + paste the configuration properties as now it would be necessary because you moved everything above. We just want him to read the configuration file and uncomment one `#` to make the configuration property active and still have that configuration property documented directly above it.

Just mention for a configuration for which it can take parameters into them, where it is not already obvious, a simple one liner is enough.

````
    # Optional parameters can be specified in the form of:
    #   parameters:
    #    param_key1: param_value1
````
this is just perfectly fine.

tiagomlalves replied:
I've reverted these changes leaving only the block above for the optional parameters.

smiklosovic commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -2925,12 +2925,12 @@ public static void setListenOnBroadcastAddress(boolean listen_on_broadcast_addre
 
     public static void setInternodeAuthenticator(IInternodeAuthenticator internodeAuthenticator)
     {
-        Preconditions.checkNotNull(internodeAuthenticator);
```
I dont understand this change. We do want to set something which is indeed non-null, no? Checking it on `get` is rather strange.

tiagomlalves replied:
@smiklosovic I'm considering reverting this change.

This change was motivated by my attempt to move the assignment of the default `IInternodeAuthenticator` from `DatabaseDescriptor` directly to `AuthConfig.applyAuth()` (aligned with the remaining authentication). 
However, this change created a cascade of other changes, namely:
 * the removal of the precondition check here because this method is called in `AuthConfig.applyAuth()` to set it for the first time (I move the precondition to the getter to ensure we would never return a null IInternodeAuthorizer).
 * fix of `ReconnectableSnitchHelperTest` which attempted to fetch `IInternodeAuthenticator` before `daemonInitialization()` was called
 * change of `clientIntialization()` to call `AuthConfig.applyAuth()` to as `IInternodeAuthenticator` is fetched from `OutboundConnection`.

The last change, further creates other issues (see [run #33](http://35.228.140.106:8080/job/cassandra/33/#showFailuresLink) ), breaking `DatabaseDescriptorRefTest` that validates the classes that are lazily initialized when calling `clientInitialization`. There are now 25 more classes that are lazily initialized by adding `AuthConfig.applyAuth()` to `clientInitialization()`. I'll push a further commit with those changes to check how fart the CI goes.

I have doubts this is really justified and I'm more inclined in simply reverting those changes. Wdyt?

smiklosovic replied:
imho just revert that

tiagomlalves replied:
I've reverted this change.

smiklosovic commented on test/data/config/version=5.0.0.yml:
```diff
@@ -126,7 +126,9 @@ tables_fail_threshold: "java.lang.Integer"
 trace_type_repair_ttl: "org.apache.cassandra.config.DurationSpec.IntSecondsBound"
 minimum_replication_factor_warn_threshold: "java.lang.Integer"
 page_size_warn_threshold: "java.lang.Integer"
-role_manager: "java.lang.String"
```
ok but ... this is not true for 5.0.0, right? We just changed something for 5.0.0.yaml.

I think 5.0-alpha1 should be just renamed to 5.0.0.yaml and then 5.0.0.yaml copied to 5.1-alpha1 and change what is different for this patch.

tiagomlalves replied:
You're right - this is not true because we have changed it for `5.0.0`. My hope was to have this change available in C* 5.0 and not in 5.1 as this is a small improvement. 

Would it be acceptable, to leave `version=5.0-alpha1.yml` unchanged and create a new `version=5.0.0.yml` with these new changes?

smiklosovic replied:
no, this patch goes to trunk which will be 5.1. This will never happen to be in any 5.0.x release.

michaelsembwever replied:
let's put this in 5.0.x

IIUC there's no compatibility break here.  it's a (small) internal operational improvement, that's permitted in patch releases.  

is it possible to keep the 5.0 yaml unchanged, and introduce the new in trunk ?   this would be in spirit of no visible user-facing change.  we have existing compatibility in place for this type of string→parameterised class change, let's take advantage of it.

IMO let's keep this PR focused on the 5.0.x patch.

smiklosovic replied:
@tiagomlalves or we can also do it for 5.0.x, right? This is just fine IMHO. Let me check it once more.

smiklosovic commented on test/unit/org/apache/cassandra/config/ConfigCompatibilityTest.java:
```diff
@@ -145,7 +151,7 @@ public void diff_4_1() throws IOException
     public void diff_5_0() throws IOException
     {
-        diff(TEST_DIR + "/version=5.0-alpha1.yml", ImmutableSet.<String>builder()
+        diff(TEST_DIR + "/version=5.0.0.yml", ImmutableSet.<String>builder()
```
@tiagomlalves so when we go to put this into 5.0.x too, just move it back to 5.0-alpha1.yaml here. Then it trunk we introduce 5.1-alpha1.yaml which will be basically a copy of 5.0-alpha1.yaml.

@michaelsembwever  does this make sense to you?

tiagomlalves replied:
I've reverted this change.

smiklosovic commented on test/unit/org/apache/cassandra/config/DatabaseDescriptorRefTest.java:
```diff
@@ -67,16 +67,29 @@ public class DatabaseDescriptorRefTest
     "org.apache.cassandra.audit.BinAuditLogger",
     "org.apache.cassandra.audit.BinLogAuditLogger",
     "org.apache.cassandra.audit.IAuditLogger",
+    "org.apache.cassandra.auth.AbstractCIDRAuthorizer",
```
@tiagomlalves  this seems like a lot. I think that by reverting that change as you suggested, we would get rid of this. We can target these changes in another ticket if you insist on changing that. I think it is too late to change it in 5.0.x, we should keep that patch as minimal as possible.

tiagomlalves replied:
I agree -  I reverted this change. My goal was to check if I could still get the CI green (which I managed). I'm not sure if there's value having these changes in another ticket.

tiagomlalves commented:
Self note: last CI Run `#35` showed only a test failure `dtest.bootstrap_test.TestBootstrap.test_shutdown_wiped_node_cannot_join` which doesn't seem to be related. Tomorrow, I'll rework the PR with the feedback.

tiagomlalves commented:
I've rebased this PR with the latest changes from `cassandra-5.0`, made changes following review comments, and re-triggered CI `#44` for this. If there are not further required change, I can squash it to a single commit with clean explanation of the changes to be merged.

tiagomlalves commented:
There are three test failures in CI `#44` - none seem to be related with these changes. I've re-triggered CI `#45` to compare. I'll double-check results tomorrow.

tiagomlalves commented:
CI `#45` shows a single test failure that was not reported in `#44`. It's safe to assume this patch didn't cause any of them.
@smiklosovic is there anything you would like to be addressed here or should I go ahead and squash all commits and rework commit message to prepare merge?

tiagomlalves commented:
@michaelsembwever and @smiklosovic ping?

smiklosovic commented:
@tiagomlalves I am tracking this, I am just busy.

smiklosovic commented:
@tiagomlalves can you please squash this and forcepush? I am getting some conflicts while doing so.

michaelsembwever commented on src/java/org/apache/cassandra/auth/AuthConfig.java:
```diff
@@ -133,4 +124,21 @@ public static void applyAuth()
         DatabaseDescriptor.getInternodeAuthenticator().validateConfiguration();
     }
+
+    public static <T> T authInstantiate(ParameterizedClass authCls, Class<T> defaultCls) {
```
can be private ?

tiagomlalves replied:
Yes. It could also be moved to `FBUtilities` as this is generic and works with any `ParameterizedClass`. I've changed the visibility but I'll postpone the refactor to `FBUtilities`.

michaelsembwever replied:
yeah, not a huge fan of polluting FBUtilities (though that is a pattern).  better to try keep things in package, even if it means duplicating small amounts of (very basic) code.

michaelsembwever commented:
LGTM.  
- one method there that looks like it can be private
- would be good to cross-check with the trunk PR, though it looks like that'll just be about bringing the _latest.yaml changes also into the main yaml
- don't forget to add summary and results ci artefacts as attachments to the jira ticket.

tiagomlalves commented:
@michaelsembwever summary and results ci artifacts were added as attachment to the jira ticket.

PR for trunk is: https://github.com/apache/cassandra/pull/3600 - it required a further change in a new C* 5.1 test `test/unit/org/apache/cassandra/transport/TlsTestUtils.java`. I'ver triggered CI `#47` for this branch.

tiagomlalves commented:
CI `#47` was aborted due to timeout. I've added 5.1 tests in trunk and fixed those of 5.0 which I believe were incorrect. I've triggered CI `#49` for those changes.

tiagomlalves commented:
CI `#49` for trunk shows 5 test failures, none of them related with these changes. 
@michaelsembwever I believe https://github.com/apache/cassandra/pull/3600 is now ready for review.

tiagomlalves committed:
CASSANDRA-19946 Standardize auth-related confs

Standardize auth-related configurations by using a short-form with
the class name instead of the full qualified class name containing
the package name.

michaelsembwever commented:
committed with https://github.com/apache/cassandra/commit/2dea5c75882ee2caaaac0ae5b84f9d6b66c46b2e
